### DISCUSSION TITLE
scalapy for mill

### DISCUSSION THREAD

**Discussion by dwalend** [com-lihaoyi/mill#1493] [Q&A]
Does anyone have an example setup for using scalapy with mill? https://scalapy.dev/docs/

**Comment by lefou** [comment]
From looking at the documentation you have linked, it should be straight forward to translate the given sbt example to Mill.

>```
> // JVM
> libraryDependencies += "me.shadaj" %% "scalapy-core" % "0.5.0"
>```

We declare a dependency on scalapy-core.

```scala
override def ivyDeps = Agg(
   ivy"me.shadaj::scalapy-core:0.5.0"
)
```

> fork := true

This is the default for `run`, `test`, `testCached`, and others. namely, all runner targets without a `Local` suffix.

>```scala
>import scala.sys.process._
>lazy val pythonLdFlags = {
>  val withoutEmbed = "python3-config --ldflags".!!
>  if (withoutEmbed.contains("-lpython")) {
>    withoutEmbed.split(' ').map(_.trim).filter(_.nonEmpty).toSeq
>  } else {
>    val withEmbed = "python3-config --ldflags --embed".!!
>    withEmbed.split(' ').map(_.trim).filter(_.nonEmpty).toSeq
>  }
>}
>
>lazy val pythonLibsDir = {
>  pythonLdFlags.find(_.startsWith("-L")).get.drop("-L".length)
>}
>```

This is more or less identical (but I prefer to use `os.proc` to execute `python3-config`) and capture the result in a target (and cache it).

```scala
def pythonLibsDir: T[String] = T{
  val pythonLdFlags = {
    val withoutEmbed = os.proc("python3-config", "--ldflags").call(cwd = T.dest).trim()
    if (withoutEmbed.contains("-lpython")) {
      withoutEmbed.split(' ').map(_.trim).filter(_.nonEmpty).toSeq
    } else {
      val withEmbed = os.proc("python3-config", "--ldflags", "--embed").call(cwd = T.dest).trim()
      withEmbed.split(' ').map(_.trim).filter(_.nonEmpty).toSeq
    }
  }
  val pythonLibsDir = {
    pythonLdFlags.find(_.startsWith("-L")).get.drop("-L".length)
  }
  pythonLibsDir
}
```

>```
>javaOptions += s"-Djna.library.path=$pythonLibsDir"
>```

Then we override `def forkArgs` to configure the `jna.library.path`, which is used by the `run` target.

```scala
override def forkArgs = Seq(s"-Djna.library.path=${pythonLibsDir()}")
```

>```
>export SCALAPY_PYTHON_LIBRARY=python3.8
>```

```scala
override def forkEnv = Map("SCALAPY_PYTHON_LIBRARY" -> "python3.8")
```

>```
>sbt run
>```

```shell
$ mill <yourmodule>.run
```

**Reply by dwalend** [reply]
Thanks! I'll give it a try this weekend.

**Comment by dev590t** [comment]
The code of lefou has some little compilation error. Here a other solution base on `python-native-libs`. It is tested.
```
import $ivy.`ai.kien::python-native-libs:0.2.1`
trait ScalaPy extends ScalaModule {
  override def ivyDeps = Agg(
    ivy"me.shadaj::scalapy-core:0.5.0"
  ) ++ super.ivyDeps()

  override def forkArgs = {
    import ai.kien.python.Python
    lazy val python = Python()
    lazy val javaOpts = python.scalapyProperties.get.map { case (k, v) =>
      s"""-D$k=$v"""
    }.toSeq
    javaOpts
  }
}

```
